edgelet/iotedge/src/check/checks/container_engine_dns.rs

```diff
@@ -0,0 +1,79 @@
+use std::fs::File;
+
+use failure::{self, Context, ResultExt};
+
+use crate::check::{checker::Checker, Check, CheckResult};
+
+#[derive(Default, serde_derive::Serialize)]
+pub(crate) struct ContainerEngineDns {
+    container_engine_config_path: Option<String>,
+    dns: Option<Vec<String>>,
+}
+
+impl Checker for ContainerEngineDns {
+    fn id(&self) -> &'static str {
+        "container-engine-dns"
+    }
+    fn description(&self) -> &'static str {
+        "DNS server"
+    }
+    fn execute(&mut self, check: &mut Check) -> CheckResult {
+        self.inner_execute(check)
+            .unwrap_or_else(CheckResult::Failed)
+    }
+    fn get_json(&self) -> serde_json::Value {
+        serde_json::to_value(self).unwrap()
+    }
+}
+
+impl ContainerEngineDns {
+    fn inner_execute(&mut self, check: &mut Check) -> Result<CheckResult, failure::Error> {
+        const MESSAGE: &str =
+        "Container engine is not configured with DNS server setting, which may impact connectivity to IoT Hub.\n\
+         Please see https://aka.ms/iotedge-prod-checklist-dns for best practices.\n\
+         You can ignore this warning if you are setting DNS server per module in the Edge deployment.";
+
+        self.container_engine_config_path = Some(
+            check
+                .container_engine_config_path
+                .to_str()
+                .unwrap_or_else(|| "No path found")
```

Comment:
`"No path found"` is not right - `to_str()` failing means `check.container_engine_config_path` is not UTF-8.

Anyway this should be `Some(check.container_engine_config_path.to_str_lossy().into_owned())`